Add basic support for measure scripts

alexdewar opened the pull request:
@jonemurray: I've tagged you as a reviewer for this PR, but don't feel any pressure to actually do anything, unless you want to! I just thought you might have an opinion on what I've done here, so I thought I'd tag you in it and give you a chance to have a look at the code, if you like.

This PR adds some basic code and documentation for what I've called "measure scripts". (I think in the old program they're  just referred to as "scripts" but I thought that was a bit ambiguous.)

The idea is that the program needs to be able to move the mirror to a sequence of positions, taking a specified number of recordings at each one. The whole process can be repeated an arbitrary number of times. We decided against keeping the old file format for these measure scripts and so I've implemented this using YAML. I've added the details to a doc in this PR. Obviously, we don't have code for controlling the actual mirror yet, so currently the code just controls a `DummyStepperMotor` by sending `"stepper.move"` messages.

I've also written code for creating and editing these files with a point-and-click interface (see #33), but mindful of discussions we've had elsewhere recently about overlong PRs, I thought it would be best to split it up :wink: so we can discuss the file format separately from the GUI code to manipulate it.

Closes #16.

dalonsoa commented on docs/measure_scripts.md:
````diff
@@ -0,0 +1,25 @@
+These files are written in [YAML](https://yaml.org/) format, with certain properties.
+Here is an example:
+
+```yaml
+measurements:
+  count: 10
+  sequence:
+  - angle: zenith
+    count: 1
+  - angle: 10.0
+    count: 3
+```
````
Why the top level `measurement`? This can be written in a more simpler form as:

```suggestion
```yaml
count: 10
sequence:
- angle: zenith
  count: 1
- angle: 10.0
  count: 3
```
```

dalonsoa replied:
Also, I would replace the top level `count` by `repeat` and the inner `count` by `measurements` (or `meas`), so their meaning is more explicit.

alexdewar replied:
In my head I had the idea that we might want to add another top-level key at some point. We probably should have a way of versioning the file format (with a `version` key or something) in case we decide to add features in future. And it sort of files like that info shouldn't be at the same level as things that are *actually* to do with how measurements will be recorded, e.g. `count` etc. What do you think?

I take your point about the other keys and I'll change those.

dalonsoa replied:
Ok. Is there any hint that other features will be necessary in a measure script in the future? If not, I would not make the nesting deeper than needed - certainly not if the only other reason is to keep `version` - which I think it is a very appropriate field - separated from the others.

alexdewar replied:
Okey doke. These scripts are only ever going to be for describing the parameters to collect measurements, so I can't see there being any more top-level keys needed. Let's just drop the `measurements` one.

I suppose it would probably be useful to have a top-level `metadata` key with the creation date, program version etc. too as well as the file format version while we're at it.

dalonsoa replied:
Having that top level metadata key with other stuff makes more sense to me than the measurement one, so go ahead with it if you want.

dalonsoa commented on tests/test_measure_script.py:
```diff
@@ -0,0 +1,84 @@
+from finesse.gui.measure_script.parse import ParseError, parse_script
+
+
+def test_valid_script() -> None:
```
As pointed out in another PR, this could be made more concise using parametric testing.

alexdewar replied:
Good shout. I wrote this before that discussion, but you're quite right! I'll do that.

dalonsoa reviewed (changes requested):
Looks good. My main comment is that I think the structure of the file should be simplified by removing the top level `measurements` section, as well as renaming some of the fields.

alexdewar commented on finesse/gui/measure_script/parse.py:
```diff
@@ -0,0 +1,91 @@
+    )
+
+    try:
+        return schema.validate(yaml.safe_load(script))["measurements"]
```
That case will be caught by the `schema.validate()`.

alexdewar commented:
Right, I've added some parametrised tests now so I think we're good to go! I added some missing checks to the code along the way that I wouldn't have spotted if I hadn't done this, so it has been a useful exercise.

dc2917 commented on finesse/gui/measure_script/__init__.py:
```diff
@@ -0,0 +1,5 @@
+"""Measure scripts represent a sequence of mirror positions for a recording.
+
+This module contains code for saving and loading mirror scripts, as well as creating and
```
mirror scripts -> measure scripts?

alexdewar replied:
Good spot!

dalonsoa replied:
Well, actually itbis both, measure and mirror scripts as they set the position pf the mirror

alexdewar replied:
True. I did toy with calling them "mirror scripts" initially, but eventually I settled on calling them "measure scripts" everywhere to avoid confusion.

dalonsoa commented on docs/measure_scripts.md:
````diff
@@ -0,0 +1,24 @@
+These files are written in [YAML](https://yaml.org/) format, with certain properties.
+Here is an example:
+
+```yaml
+count: 10
+sequence:
+- angle: zenith
+  count: 1
+- angle: 10.0
+  count: 3
+```
````
I thought we have agreed to give a more informative name to the `count` keyword in the two contexts. See https://github.com/ImperialCollegeLondon/FINESSE/pull/63#discussion_r1043039325

alexdewar replied:
My bad. I'll do that now

dalonsoa commented on finesse/gui/measure_script/parse.py:
```diff
@@ -0,0 +1,115 @@
+    def run(self) -> None:
+        """Run this measure script."""
+        logging.info(f"Running {self.path}")
+        for i in range(self.count):
```
I'd use [`tqdm`](https://pypi.org/project/tqdm/) here. Eventually, we will want a GUI-based progress bar somewhere that also allows to cancel the sequence, but that will be work for the future, when the basic stuff is up and running. Something like [QProgressDialog](https://doc.qt.io/qtforpython-5/PySide2/QtWidgets/QProgressDialog.html). Maybe open an issue about it, so we do not forget.

alexdewar replied:
This is a neat package. I knocked something together using it, but the problem is that it obscures the console output and atm that's the only acknowledgement we get that the mirror has been "moved" to a given position, so I think we should drop it for now.

I've opened an issue: #71

dalonsoa reviewed (changes requested):
I"ve a couple of comments, but it looks OK, overall.

dalonsoa commented on finesse/gui/measure_script/parse.py:
```diff
@@ -0,0 +1,120 @@
+
+
+@dataclass
+class Measurement:
+    """Represents a single measurement (i.e. angle + number of measurements)."""
```
Should this be called `Step` rather than `Measurement` to avoid mixing it with the individual acquisitions?

alexdewar replied:
Good point.

dalonsoa reviewed (approved):
Looks good! I've just made a comment.
